compiler/qsc_linter/src/lints/hir.rs

```diff
@@ -29,7 +35,10 @@ use super::lint;
 //  For more details on how to add a lint, please refer to the crate-level documentation
 //  in qsc_linter/lib.rs.
 declare_hir_lints! {
-    (NeedlessOperation, LintLevel::Allow, "operation does not contain any quantum operations", "this callable can be declared as a function instead")
+    (NeedlessOperation, LintLevel::Allow, "operation does not contain any quantum operations", "this callable can be declared as a function instead"),
+    (DeprecatedFunctionConstructor, LintLevel::Warn, "deprecated function constructors", "function constructors for struct types are deprecated, use `new` instead"),
+    (DeprecatedWithOperator, LintLevel::Warn, "deprecated `w/` and `w/=` operators for structs", "`w/` and `w/=` operators for structs are deprecated, use `new` instead"),
+    (DeprecatedDoubleColonOperator, LintLevel::Warn, "deprecated `::` for field access", "`::` operator is deprecated, use `.` instead"),
```

Comment:
Admittedly I haven't looked into the implementation yet, but in general, why do these need to be HIR lints? My first thought was that these are syntactic concerns so you should only need the AST?

Reply:
The new lints all use the compilation to do lookups for types, name references, or source code. The AST lints don't make use of the compilation, so it is not plumbed through the AST lints.